### ISSUE TITLE
Migrate all external calls to a GNOME wrapper module

### ISSUE THREAD

**Issue by mmstick** [pop-os/shell#109]
All calls into GNOME APIs should be routed through a wrapper (`gnome.ts`). This is purely a QoL improvement that will make the project easier to maintain. This issue is for interested contributors to the Pop Shell. The work involved is primarily that of refactoring. 

By keeping all external API calls in a single module, this will make it easier to quickly adjust to new changes in the GNOME Shell API over time. So instead of having to update dozens of call sites to fix a breaking change in GNOME, we'd only need to update the wrapper.

This also has the benefit of giving us stronger guarantees over types in GNOME functions. Please declare all types involved in each function, as well as interfaces and classes involved.

See [GJS Docs](https://gjs-docs.gnome.org) for API documentation in GJS. You should enable Clutter, Meta, Shell, St, GLib, GObject, and Gio. Undocumented classes can be found in the [GNOME Shell source code](https://github.com/GNOME/gnome-shell/tree/master/js).

**Comment by mcaneris** [comment]
Hey @mmstick, as a partial example, is this what you have in mind?

```
// gnome.ts

const { GLib } = imports.gi;

interface GLibSourceFunction {
  (): boolean;
}

export class GLibWrapper {
  public static PRIORITY_LOW: number;
  public static PRIORITY_DEFAULT: number;

  public static idle_add(priority: number, fn: GLibSourceFunction): number {
    return GLib.idle_add(priority, fn);
  }

  public static timeout_add(
    priority: number,
    interval: number,
    fn: GLibSourceFunction
  ): number {
    return GLib.timeout_add(priority, interval, fn);
  }

  public static get_home_dir(): string {
    return GLib.get_home_dir();
  }
}
```

```
// extension.ts
import * as gnome from 'gnome';
const { GLibWrapper: GLib } = gnome;
```

I would like to work on this if that's ok, I would appreciate any guidance you can provide.

**Comment by mmstick** [comment]
Yes. All calls to GNOME APIs should be carried out through this module.

First, there should be a declare for each of the classes from GNOME APIs. I already have a collection of declares for GNOME in the `mod.d.ts` file, which we should rename to `gnome.d.ts`.

Then, we should build more convenient functions and classes on top of that that expose the functionality and handle any sort of quirks. If you know that a function has a possibility of throwing an exception, it'd be ideal if we could catch that and make the return type `Result<T, E>`, using the `Result` class in the `result` module, and `Error` class in the `error` module. In fact, any function that can error should return a `Result` to indicate the reason for error.

Only the GNOME module should be calling into `imports`, and you may see some patterns in GNOME API usage that are best moved to this wrapper. LIkewise, if you find any classes that are holding GNOME classes, they should be converted to holding wrappers of GNOME classes.

Instead of 
```
import * as gnome from 'gnome';
const { GLibWrapper: GLib } = gnome;
```

It should be

```
import * as gnome from 'gnome';
const { GLib } = gnome;
```

If you have any ideas of ways to improve the existing code, that would also be great.

**Comment by mcaneris** [comment]
Hey @mmstick, what do you think about [this partial implementation](https://github.com/mcaneris/shell/tree/wrap_gnome)? I wanted to confirm with you before I go deeper into the rabbit hole.

- slowly moving definitions from mod.d.ts to gnome.d.ts to ensure it keeps compiling.
- moved definitions are under Gnome namespace to prevent collision with wrapped ones.
- keeping scopes like gi, misc in exports.
- calls pass through directly to Gnome libs unless we wrap.
- there is a Result<T, E> wrap implementation in xprop.ts.
- it compiles :).

I'm not sure on how to deal with the getCurrentExtension() call requirement, any ideas?:
```
// @ts-ignore
const Me = imports.misc.extensionUtils.getCurrentExtension();
```

**Comment by scaryrawr** [comment]
It looks like there's some tools to generate declaration files [ts-for-gjs](https://github.com/sammydre/ts-for-gjs) there's other ones as well. I haven't tried it though.
